### PR TITLE
fix(web-components): fix disabled button styles

### LINKED ISSUE #31552: Disabled styles on subtle buttons is not applied correctly
https://github.com/microsoft/fluentui/blob/80e5bbdcf7dd54ceec4050d71edd6089d42b4159/packages/web-components/src/button/button.styles.ts#L258

Using disabled subtle styled buttons applies the wrong styling. Background and border should be transparent, but in the wild they get `colorNeutralBackgroundDisabled` and `colorNeutralStrokeDisabled`. This issue is live on the [storybook site](https://web-components.fluentui.dev/?path=/docs/components-button-button--button).

### PR DESCRIPTION
**Pull request by davatron5000**
## Previous Behavior

- Bug filed for `[disabled][appearance="subtle"]` fluent-buttons where they are expected to not have a background or outline when disabled, but actually were showing a background and outline.
  - Confirmed that a "fix" in https://github.com/microsoft/fluentui/commit/fa9416a9f3fefc36b2e2ee3b32ea8c038bb4d91e caused a regression where "fixed" styles were now overriding existing behavior due to cascade order.
  - This exposed a false positive in our Playwright tests where CSS variables were being wiped out by the use of `setContent`.
- Discovered a "`focusabale`" typo in button styles
- Discovered a bug with `:focus-visible` not working in High Contrast Mode

## New Behavior

- Reorganized CSS so that `[disabled]` styles cascade correctly.
- Fixed `focusabale` typo
- Removed errant `)` char from `:focus-visible` styles to fix high-contrast mode focus states.
- Thanks to help from @radium-v and prior investigation by @janechu, changed how we register CSS properties in `setTheme` to prevent them being wiped out by `page.setContent`
  - Now using `CSS.registerProperty` with a fallback to current method for current versions of Firefox and older browsers.
- Fixed and improved tests to help preserve behavior.

The `CSS.registerProperty` change is probably significant enough to heavily scrutinize this PR. It's pretty naive at the moment, just applying an `initialValue`. In the future I think we'd want to set actual CSSOM types (integer, pixel values, etc). 

Look forward to feedback.

## Related Issue(s)

- Fixes #31552 

## Screenshots 

Screenshot 2024-06-05 at 11 34 42 AM

Screenshot 2024-06-05 at 11 36 00 AM

### CODE REVIEW

File: packages/web-components/src/theme/set-theme.ts [hunk 1]
```diff
@@ -10,7 +10,15 @@ const tokenNames = Object.keys(tokens) as (keyof Theme)[];
...
-    document.body.style.setProperty(`--${t}`, theme[t] as string);
+    if ('registerProperty' in CSS) {
+      CSS.registerProperty({
+        name: `--${t}`,
+        inherits: true,
+        initialValue: theme[t] as string,
+      });
+    } else {
+      document.body.style.setProperty(`--${t}`, theme[t] as string);
+    }
```
Comment (chrisdholt) [inline packages/web-components/src/theme/set-theme.ts:21]: Just curious - do we know the performance delta here?

Comment (davatron5000) [reply packages/web-components/src/theme/set-theme.ts:21]: I don't know off-hand. But let me see if I can get you an answer.

File: change/@fluentui-web-components-3a27963c-2795-4898-8b1e-adeb72be5c72.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-web-components-3a27963c-2795-4898-8b1e-adeb72be5c72.json:1]: ## [🕵 fluentui-web-components-v3 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31585&commitId=19517c4fb45443673c6dd638e961883fc467aa93&env=prod&clientType=FLUENTUI)


  <details><summary><b>Accordion</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Accordion.Size.normal.chromium.png|2| Changed|
|Accordion.Size - Dark Mode.normal.chromium.png|2| Changed|
|Accordion.Size - RTL.normal.chromium.png|2| Changed|

</details>


  <details><summary><b>Avatar</b> 5 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar.Active Appearance.normal.chromium.png|10| Changed|
|Avatar.Active Appearance - Dark Mode.normal.chromium.png|2| Changed|
|Avatar.Color - RTL.normal.chromium.png|1| Changed|
|Avatar.Active.normal.chromium.png|5| Changed|
|Avatar.Color.normal.chromium.png|1| Changed|

</details>


  <details><summary><b>Badge</b> 6 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Badge.Appearance - Dark Mode.normal.chromium.png|15| Changed|
|Badge.Size.normal.chromium.png|6| Changed|
|Badge.Appearance.normal.chromium.png|1| Changed|
|Badge.Default.normal.chromium.png|1| Changed|
|Badge.Shape.normal.chromium.png|1| Changed|
|Badge.Color.normal.chromium.png|1| Changed|

</details>


  <details><summary><b>Button</b> 57 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Button.Large Icon Only - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large Icon Only.pressed.chromium.png|3| Changed|
|Button.Large With Icon - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large With Icon - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large Icon Only.hover.chromium.png|4| Changed|
|Button.Large With Icon - RTL.default.chromium.png|5| Changed|
|Button.Large Icon Only - Dark Mode.default.chromium.png|4| Changed|
|Button.Large With Icon.hover.chromium.png|4| Changed|
|Button.Large Icon Only.default.chromium.png|5| Changed|
|Button.Large With Icon.pressed.chromium.png|3| Changed|
|Button.Large With Icon - RTL.hover.chromium.png|4| Changed|
|Button.Large With Icon - Dark Mode.default.chromium.png|4| Changed|
|Button.Large With Icon.default.chromium.png|5| Changed|
|Button.Large Icon Only - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large With Icon - RTL.pressed.chromium.png|3| Changed|
|Button.Outline Disabled Focusable - Dark Mode.default.chromium.png|159| Changed|
|Button.Outline Disabled - Dark Mode.pressed.chromium.png|159| Changed|
|Button.Primary Disabled - Dark Mode.default.chromium.png|44| Changed|
|Button.Outline Disabled Focusable - Dark Mode.pressed.chromium.png|159| Changed|
|Button.Primary Disabled - Dark Mode.pressed.chromium.png|44| Changed|
|Button.Primary Disabled Focusable - Dark Mode.default.chromium.png|44| Changed|
|Button.Primary Disabled Focusable - Dark Mode.pressed.chromium.png|44| Changed|
|Button.Primary Disabled.hover.chromium.png|15| Changed|
|Button.Primary Disabled - Dark Mode.hover.chromium.png|44| Changed|
|Button.Primary Disabled.default.chromium.png|15| Changed|
|Button.Primary Disabled.pressed.chromium.png|15| Changed|
|Button.Outline Disabled - Dark Mode.default.chromium.png|159| Changed|
|Button.Outline Disabled Focusable - Dark Mode.hover.chromium.png|159| Changed|
|Button.Primary Disabled Focusable.hover.chromium.png|15| Changed|
|Button.Outline Disabled - Dark Mode.hover.chromium.png|159| Changed|
|Button.Primary Disabled Focusable - Dark Mode.hover.chromium.png|44| Changed|
|Button.Subtle Disabled - Dark Mode.pressed.chromium.png|187| Changed|
|Button.Subtle Disabled - Dark Mode.default.chromium.png|187| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.pressed.chromium.png|187| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.hover.chromium.png|187| Changed|
|Button.Subtle Disabled Focusable.hover.chromium.png|25| Changed|
|Button.Transparent Disabled - Dark Mode.default.chromium.png|199| Changed|
|Button.Primary Disabled Focusable.default.chromium.png|15| Changed|
|Button.Transparent Disabled - Dark Mode.pressed.chromium.png|199| Changed|
|Button.Primary Disabled Focusable.pressed.chromium.png|15| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.pressed.chromium.png|199| Changed|
|Button.Transparent Disabled Focusable.pressed.chromium.png|27| Changed|
|Button.Subtle Disabled - Dark Mode.hover.chromium.png|187| Changed|
|Button.Transparent Disabled.hover.chromium.png|27| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.default.chromium.png|187| Changed|
|Button.Transparent Disabled.pressed.chromium.png|27| Changed|
|Button.Subtle Disabled.pressed.chromium.png|25| Changed|
|Button.Transparent Disabled Focusable.hover.chromium.png|27| Changed|
|Button.Subtle Disabled.hover.chromium.png|25| Changed|
|Button.Transparent Disabled Focusable.default.chromium.png|27| Changed|
|Button.Subtle Disabled Focusable.pressed.chromium.png|25| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.default.chromium.png|199| Changed|
|Button.Subtle Disabled Focusable.default.chromium.png|25| Changed|
|Button.Subtle Disabled.default.chromium.png|25| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.hover.chromium.png|199| Changed|
|Button.Transparent Disabled.default.chromium.png|27| Changed|
|Button.Transparent Disabled - Dark Mode.hover.chromium.png|199| Changed|

</details>


  <details><summary><b>Divider</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Divider.With Svg - RTL.normal.chromium.png|3| Changed|
|Divider.With Svg.normal.chromium.png|3| Changed|
|Divider.Vertical With Svg.normal.chromium.png|4| Changed|

</details>


  <details><summary><b>MenuList</b> 16 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|MenuList.Checkbox - RTL.2nd selected.chromium.png|1| Changed|
|MenuList.Checkbox With Icons.2nd selected.chromium.png|5| Changed|
|MenuList.Checkbox With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons.normal.chromium.png|4| Changed|
|MenuList.Checkbox.2nd selected.chromium.png|1| Changed|
|MenuList.Checkbox With Icons.1st selected.chromium.png|4| Changed|
|MenuList.Radio With Icons.2nd selected.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.2nd selected.chromium.png|5| Changed|
|MenuList.Radio With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.2nd selected.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.Radio With Icons.normal.chromium.png|4| Changed|
|MenuList.Radio With Icons.1st selected.chromium.png|4| Changed|
|MenuList.With Icons.hover menuitem.chromium.png|3| Changed|
|MenuList.With Icons.normal.chromium.png|4| Changed|

</details>


  <details><summary><b>Slider</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Slider.Size Small.rightArrow.chromium.png|1| Changed|
|Slider.Size Small.normal.chromium.png|1| Changed|

</details>


  <details><summary><b>Switch</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Switch.Checked - RTL.hover.chromium.png|1| Changed|
|Switch.Checked.hover.chromium.png|1| Changed|

</details>


  <details><summary><b>Text</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Text.Block.normal.chromium.png|995| Changed|

</details>


  <details><summary><b>TextInput</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|TextInput.Size Large.normal.chromium.png|2| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_web-components"/>

File: packages/web-components/src/theme/set-theme.ts (hunk 1 above)
Comment (davatron5000) [reply packages/web-components/src/theme/set-theme.ts:21]: Alright! Some initial data from a small scale test. Feel free to poke holes, or ask questions.

**Methodology**

I wrapped the function in `performance.now` calls and reported the result in the browser console, commenting out the other CSS variable registration method to get before and after timestamps. 

```js 
export const setTheme = (theme: Theme) => {
  const start = performance.now();
  for (const t of tokenNames) {
      CSS.registerProperty({
        name: `--${t}`,
        inherits: true,
        initialValue: theme[t] as string,
      });

     // document.body.style.setProperty(`--${t}`, theme[t] as string);
  }
  const end = performance.now();
  console.log(`setTheme took ${end - start}ms`);
};
```

I then ran the test at 3 different synthetic CPU profiles provided by DevTools: none, 4x, and 6x slowdown.

**Results**

| Method | No slowdown | 4x slowdown | 6x slowdown |
| --- | --- | --- | --- |
| `CSS.registerProperty` | 0.5ms | 1.8999999ms | 3ms |
| `style.setProperty` | 0.6999999ms | 2.5ms | 5.6999999ms |

The initial results of this small scale test shows a minor improvement on fast devices but a 2x improvement (but only 3ms difference) on slower devices.

Comment (chrisdholt) [reply packages/web-components/src/theme/set-theme.ts:21]: Awesome

**Review by chrisdholt — APPROVED** [review]

**Review by janechu — APPROVED** [review]

**Pull request merged**